### ISSUE TITLE
Metadata declaration design

### ISSUE THREAD

**Issue by DavidPoliakoff** [kokkos/kokkos#3725]
Looking for a design for metadata declaration. We were going to start from looking at print_configuration, which is a set of methods that looks like this:

```c++
  void print_configuration(std::ostream& out, const bool detail) {
    std::ostringstream msg;    msg << "Kokkos Version:" << std::endl;
    msg << "  " << KOKKOS_VERSION / 10000 << "." << (KOKKOS_VERSION % 10000) / 100
        << "." << KOKKOS_VERSION % 100 << std::endl;
//..... more of this form
```

Two designs and a strawman come to mind.

    Add a bool to this function to allow it to skip printing to screen. Add calls to this function that declare metadata based on what we output
    Add a separate function
    Some horrible ostreamstring contraption that declares metadata as well as building a string (sounds dumb to me, but might inspire a smarter idea)

Drawbacks of 1 are that it's a bit of a shit design for "print_configuration" to have a mode that doesn't print, and mode flags are generally awful.Drawbacks of 2 are that we're doing something mechanically similar in two places, and need to remember when we add something to "print" to add it to "declare." Also, we need to extend the interface of externally declared ExecutionSpaces to add that function.3 is just dumb, I mention it only to spur other ideas

Any preferences?

**Comment by DavidPoliakoff** [comment]
I've implemented a strawman in #3729

**Comment by DavidPoliakoff** [comment]
Closing as #3729 merged
